File: SYPictureMetadataExample/SYPictureMetadataExample/SYImagesVC.m
//
//  SYImagesVC.m
//  SYPictureMetadataExample
//
//  Created by Stanislas Chevallier on 06/02/2017.
//  Copyright © 2017 Syan. All rights reserved.
//

#import "SYImagesVC.h"
#import "SYImageVC.h"

@interface SYImagesVC () <UIPageViewControllerDataSource, UIPageViewControllerDelegate>
@property (nonatomic, assign) NSUInteger nextIndex;
@property (nonatomic, assign) NSUInteger currentIndex;
@property (nonatomic, strong) NSArray <NSURL *> *URLs;
@end

@implementation SYImagesVC

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self setDataSource:self];
    [self setDelegate:self];
    
    [self.navigationController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName:[UIColor whiteColor]}];
    [self.navigationController.navigationBar setBarTintColor:[UIColor darkGrayColor]];
    
    NSString *rootPath = [[NSBundle mainBundle] pathForResource:@"Test images" ofType:nil];
    NSArray <NSString *> *imageNames = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:rootPath error:NULL];
    
    NSMutableArray <NSURL *> *URls = [NSMutableArray array];
    for (NSString *imageName in imageNames)
    {
        NSURL *picUrl = [NSURL fileURLWithPath:[rootPath stringByAppendingPathComponent:imageName]];
        [URls addObject:picUrl];
    }
    self.URLs = [URls copy];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self openIndex:0 animated:NO forceForwardAnimation:NO];
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleLightContent;
}

#pragma mark - Data

- (void)setURLs:(NSArray<NSURL *> *)URLs
{
    self->_URLs = URLs;
    [self openIndex:self.currentIndex animated:NO forceForwardAnimation:NO];
}

- (SYImageVC *)viewControllerAtIndex:(NSUInteger)index
{
    if (!self.URLs.count)
        return nil;
    
    if (index >= self.URLs.count)
        return nil;
    
    NSURL *url = self.URLs[index];
    
    SYImageVC *vc = [self.storyboard instantiateViewControllerWithIdentifier:NSStringFromClass(SYImageVC.class)];
    [vc setPictureURL:url];
    [vc setIndex:index];
    
    return vc;
}

#pragma mark - Actions

- (void)openIndex:(NSUInteger)index animated:(BOOL)animated forceForwardAnimation:(BOOL)forceForwardAnimation
{
    UIViewController *vc = [self viewControllerAtIndex:index];
    if (!vc)
        return;
    
    UIPageViewControllerNavigationDirection direction = UIPageViewControllerNavigationDirectionForward;
    if (index < self.currentIndex && !forceForwardAnimation) // going backwards
        direction = UIPageViewControllerNavigationDirectionReverse;
    
    self.currentIndex = index;
    [self setViewControllers:@[vc] direction:direction animated:animated completion:nil];
}

- (void)setCurrentIndex:(NSUInteger)currentIndex
{
    self->_currentIndex = currentIndex;
    [self.navigationItem setTitle:self.URLs[currentIndex].lastPathComponent];
}

#pragma mark - PageController Data Source

- (nullable UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController
{
    // disable infinity scroll when only has a single item
    if (self.URLs.count == 1)
        return nil;
    
    SYImageVC *prevVC = (SYImageVC *)viewController;
    return [self viewControllerAtIndex:prevVC.index - 1];
}

- (nullable UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController
{
    // disable infinity scroll when only has a single item
    if (self.URLs.count == 1)
        return nil;
    
    SYImageVC *prevVC = (SYImageVC *)viewController;
    return [self viewControllerAtIndex:prevVC.index + 1];
}

- (void)pageViewController:(UIPageViewController *)pageViewController willTransitionToViewControllers:(NSArray<UIViewController *> *)pendingViewControllers
{
    SYImageVC *vc = (SYImageVC *)[pendingViewControllers firstObject];
    self.nextIndex = vc.index;
}

- (void)pageViewController:(UIPageViewController *)pageViewController
        didFinishAnimating:(BOOL)finished
   previousViewControllers:(NSArray <UIViewController *> *)previousViewControllers
       transitionCompleted:(BOOL)completed
{
    if (completed)
        self.currentIndex = self.nextIndex;
    
    self.nextIndex = 0;
}
@end

File: Examples/SYPictureMetadataExample/ViewControllers/SYImagesVC.swift
//
//  SYImagesVC.swift
//  SYPictureMetadataExample
//
//  Created by Stanislas Chevallier on 09/02/2020.
//  Copyright © 2020 Syan.me. All rights reserved.
//

import UIKit
import SYPictureMetadata
import SYPictureMetadataTestAssets

class SYImagesVC : UIPageViewController {
    // MARK: UIViewController
    override func viewDidLoad() {
        super.viewDidLoad()
        dataSource = self
        delegate = self
        loadURLs()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        openIndex(0, animated: false, forceForwardAnimation: false)
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
    
    // MARK: Properties
    private var nextIndex: Int = 0
    private var currentIndex: Int = 0 {
        didSet {
            navigationItem.title = urls[currentIndex].lastPathComponent
        }
    }
    private var urls: [URL] = [] {
        didSet {
            openIndex(currentIndex, animated: false, forceForwardAnimation: false)
        }
    }
    
    // MARK: Content
    private func loadURLs() {
        self.urls = TestFile.allCases.map(\.url)
    }
    
    // MARK: Child ViewControllers
    private func viewController(at newIndex: Int) -> SYImageVC? {
        guard urls.count > 0 else { return nil }
        let index = (newIndex + urls.count) % urls.count
        guard index < urls.count else { return nil }
        let url = urls[index]
        
        let vc = SYImageVC()
        vc.pictureURL = url
        vc.index = index
        
        return vc
    }
    
    private func openIndex(_ index: Int, animated: Bool, forceForwardAnimation: Bool) {
        guard let vc = viewController(at: index) else { return }
        
        var direction = UIPageViewController.NavigationDirection.forward
        if (index < currentIndex && !forceForwardAnimation) {
            // going backwards
            direction = .reverse
        }

        currentIndex = index
        setViewControllers([vc], direction: direction, animated: animated, completion: nil)
    }
}

extension SYImagesVC : UIPageViewControllerDataSource {
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        // disable infinity scroll when only has a single item
        if urls.count < 2 { return nil }
        
        guard let prevVC = viewController as? SYImageVC else { return nil }
        return self.viewController(at: prevVC.index - 1)
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        // disable infinity scroll when only has a single item
        if urls.count < 2 { return nil }
        
        guard let prevVC = viewController as? SYImageVC else { return nil }
        return self.viewController(at: prevVC.index + 1)
    }
}

extension SYImagesVC : UIPageViewControllerDelegate {
    func pageViewController(_ pageViewController: UIPageViewController, willTransitionTo pendingViewControllers: [UIViewController]) {
        if let vc = pendingViewControllers.first as? SYImageVC {
            nextIndex = vc.index
        }
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
        if completed {
            currentIndex = nextIndex
        }
        nextIndex = 0
    }
}

